Panic because of invalid result during xDS translation

*Description*:
When integrating Envoy Gateway and Envoy Gateway Extension service, I observed that during the xDS translation result, when an error occurred, it would be panicking because there was some operation to the result variable before the error check.
https://github.com/envoyproxy/gateway/blob/7d9e16e4997e7118d70ea910a6ef9d0732426878/internal/xds/translator/runner/runner.go#L80-L98

*Repro steps*:

In my scenario, we can simply produce an error from one of the hooks registered through the Envoy Gateway Extension service. Then you will see Envoy Gateway no longer up and running because of panicking.

*Environment*:
Tested on Envoy Gateway v0.5.0, but the code also persists in the main branch.

*Logs*:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1792a64]

goroutine 97 [running]:
github.com/envoyproxy/gateway/internal/xds/translator/runner.(*Runner).subscribeAndTranslate.func1({{0x400086d170?, 0x1733654?}, 0x78?, 0x4000138f50?})
        /home/runner/work/gateway/gateway/internal/xds/translator/runner/runner.go:79 +0x1a4
github.com/envoyproxy/gateway/internal/message.HandleSubscription[...](0x2112300, 0x4000d4df88?)
        /home/runner/work/gateway/gateway/internal/message/watchutil.go:36 +0x104
github.com/envoyproxy/gateway/internal/xds/translator/runner.(*Runner).subscribeAndTranslate(0x40003ba180, {0x20f19b8?, 0x4000528410?})
        /home/runner/work/gateway/gateway/internal/xds/translator/runner/runner.go:52 +0x70
created by github.com/envoyproxy/gateway/internal/xds/translator/runner.(*Runner).Start
        /home/runner/work/gateway/gateway/internal/xds/translator/runner/runner.go:45 +0x1f8
```

*Expectation*:

The error during xDS translation should be returned and logged instead of panicking.

--- a/internal/xds/translator/runner/runner.go
+++ b/internal/xds/translator/runner/runner.go
@@ -78,6 +78,10 @@ func (r *Runner) subscribeAndTranslate(ctx context.Context) {
 				}
 
 				result, err := t.Translate(val)
+				if err != nil {
+					r.Logger.Error(err, "failed to translate xds ir")
+					return
+				}
 
 				// Publish EnvoyPatchPolicyStatus
 				for _, e := range result.EnvoyPatchPolicyStatuses {
@@ -90,12 +94,8 @@ func (r *Runner) subscribeAndTranslate(ctx context.Context) {
 				// Discard the EnvoyPatchPolicyStatuses to reduce memory footprint
 				result.EnvoyPatchPolicyStatuses = nil
 
-				if err != nil {
-					r.Logger.Error(err, "failed to translate xds ir")
-				} else {
-					// Publish
-					r.Xds.Store(key, result)
-				}
+				// Publish
+				r.Xds.Store(key, result)
 			}
 		},
 	)

--- a/internal/xds/translator/runner/runner.go
+++ b/internal/xds/translator/runner/runner.go
@@ -80,22 +80,21 @@ func (r *Runner) subscribeAndTranslate(ctx context.Context) {
 				result, err := t.Translate(val)
 				if err != nil {
 					r.Logger.Error(err, "failed to translate xds ir")
-					return
-				}
-
-				// Publish EnvoyPatchPolicyStatus
-				for _, e := range result.EnvoyPatchPolicyStatuses {
-					key := ktypes.NamespacedName{
-						Name:      e.Name,
-						Namespace: e.Namespace,
+				} else {
+					// Publish EnvoyPatchPolicyStatus
+					for _, e := range result.EnvoyPatchPolicyStatuses {
+						key := ktypes.NamespacedName{
+							Name:      e.Name,
+							Namespace: e.Namespace,
+						}
+						r.EnvoyPatchPolicyStatuses.Store(key, e.Status)
 					}
-					r.EnvoyPatchPolicyStatuses.Store(key, e.Status)
-				}
-				// Discard the EnvoyPatchPolicyStatuses to reduce memory footprint
-				result.EnvoyPatchPolicyStatuses = nil
+					// Discard the EnvoyPatchPolicyStatuses to reduce memory footprint
+					result.EnvoyPatchPolicyStatuses = nil
 
-				// Publish
-				r.Xds.Store(key, result)
+					// Publish
+					r.Xds.Store(key, result)
+				}
 			}
 		},
 	)

--- a/internal/xds/translator/runner/runner.go
+++ b/internal/xds/translator/runner/runner.go
@@ -80,21 +80,22 @@ func (r *Runner) subscribeAndTranslate(ctx context.Context) {
 				result, err := t.Translate(val)
 				if err != nil {
 					r.Logger.Error(err, "failed to translate xds ir")
-				} else {
-					// Publish EnvoyPatchPolicyStatus
-					for _, e := range result.EnvoyPatchPolicyStatuses {
-						key := ktypes.NamespacedName{
-							Name:      e.Name,
-							Namespace: e.Namespace,
-						}
-						r.EnvoyPatchPolicyStatuses.Store(key, e.Status)
-					}
-					// Discard the EnvoyPatchPolicyStatuses to reduce memory footprint
-					result.EnvoyPatchPolicyStatuses = nil
-
-					// Publish
-					r.Xds.Store(key, result)
+					return
 				}
+
+				// Publish EnvoyPatchPolicyStatus
+				for _, e := range result.EnvoyPatchPolicyStatuses {
+					key := ktypes.NamespacedName{
+						Name:      e.Name,
+						Namespace: e.Namespace,
+					}
+					r.EnvoyPatchPolicyStatuses.Store(key, e.Status)
+				}
+				// Discard the EnvoyPatchPolicyStatuses to reduce memory footprint
+				result.EnvoyPatchPolicyStatuses = nil
+
+				// Publish
+				r.Xds.Store(key, result)
 			}
 		},
 	)

--- a/internal/xds/translator/runner/runner_test.go
+++ b/internal/xds/translator/runner/runner_test.go
@@ -7,13 +7,18 @@ package runner
 
 import (
 	"context"
+	"fmt"
 	"testing"
 	"time"
 
+	listenerv3 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
 	resourcev3 "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+
 	"github.com/stretchr/testify/require"
 
+	"github.com/envoyproxy/gateway/api/v1alpha1"
 	"github.com/envoyproxy/gateway/internal/envoygateway/config"
+	"github.com/envoyproxy/gateway/internal/extension/types"
 	"github.com/envoyproxy/gateway/internal/ir"
 	"github.com/envoyproxy/gateway/internal/message"
 )
@@ -92,4 +97,86 @@ func TestRunner(t *testing.T) {
 		return len(out) == 0
 	}, time.Second*5, time.Millisecond*50)
 
+}
+
+func TestRunner_withExtensionManager(t *testing.T) {
+	// Setup
+	xdsIR := new(message.XdsIR)
+	xds := new(message.Xds)
+	cfg, err := config.New()
+	require.NoError(t, err)
+	r := New(&Config{
+		Server:           *cfg,
+		XdsIR:            xdsIR,
+		Xds:              xds,
+		ExtensionManager: &extManagerMock{},
+	})
+
+	ctx := context.Background()
+	// Start
+	err = r.Start(ctx)
+	require.NoError(t, err)
+
+	// xDS is nil at start
+	require.Equal(t, map[string]*ir.Xds{}, xdsIR.LoadAll())
+
+	// test translation
+	path := "example"
+	res := ir.Xds{
+		HTTP: []*ir.HTTPListener{
+			{
+				Name:      "test",
+				Address:   "0.0.0.0",
+				Port:      80,
+				Hostnames: []string{"example.com"},
+				Routes: []*ir.HTTPRoute{
+					{
+						Name: "test-route",
+						PathMatch: &ir.StringMatch{
+							Exact: &path,
+						},
+						Destination: &ir.RouteDestination{
+							Name: "test-dest",
+							Settings: []*ir.DestinationSetting{
+								{
+									Endpoints: []*ir.DestinationEndpoint{
+										{
+											Host: "10.11.12.13",
+											Port: 8080,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	xdsIR.Store("test", &res)
+	require.Eventually(t, func() bool {
+		out := xds.LoadAll()
+		// Ensure that xds has no key, value pairs
+		return len(out) == 0
+	}, time.Second*5, time.Millisecond*50)
+}
+
+type extManagerMock struct {
+	types.Manager
+}
+
+func (m *extManagerMock) GetPostXDSHookClient(xdsHookType v1alpha1.XDSTranslatorHook) types.XDSHookClient {
+	if xdsHookType == v1alpha1.XDSHTTPListener {
+		return &xdsHookClientMock{}
+	}
+
+	return nil
+}
+
+type xdsHookClientMock struct {
+	types.XDSHookClient
+}
+
+func (c *xdsHookClientMock) PostHTTPListenerModifyHook(*listenerv3.Listener) (*listenerv3.Listener, error) {
+	return nil, fmt.Errorf("assuming a network error during the call")
 }

